### ISSUE TITLE
Can this daemon be compiled with windows

### ISSUE THREAD

**Issue by manojc99** [GENIVI/dlt-daemon#136]
Is it possible to compile this with windows and create a dll instead of so file.

**Comment by ssugiura** [comment]
@manojc99 I'm afraid that current DLT cannot be compiled on Windows. It seems initial version of dlt-daemon supported Windows, but after that it was considered for Linux usage only, so some headers, which are not available on Windows, are not excluded for Windows.

**Comment by ssugiura** [comment]
@manojc99 Do you have any updates from your side?

**Comment by chenlimingcn** [comment]
you can comile by using cygwin on windows, I have done

**Comment by manojc99** [comment]
Thanks for the reply Rimon. Any instructions that you can share please

**Comment by chenlimingcn** [comment]
git pull to this version eb3dcd26c80c4ac6e431b09b50b0861b9753ceea(master)
and then modified as follow diff output

diff --git a/src/daemon/dlt-daemon.c b/src/daemon/dlt-daemon.c
index b7766f0..ed548f0 100644
--- a/src/daemon/dlt-daemon.c
+++ b/src/daemon/dlt-daemon.c
@@ -1062,14 +1062,14 @@ static int dlt_daemon_init_fifo(DltDaemonLocal *daemon_local)
 
     if (daemon_local->daemonFifoSize != 0) {
         /* Set Daemon FIFO size */
-        if (fcntl(fd, F_SETPIPE_SZ, daemon_local->daemonFifoSize) == -1) {
+        if (fcntl(fd, FD_SETSIZE, daemon_local->daemonFifoSize) == -1) {
             snprintf(str, DLT_DAEMON_TEXTBUFSIZE, "set FIFO size error: %s\n", strerror(errno));
             dlt_log(LOG_ERR, str);
         }
     }
 
     /* Get Daemon FIFO size */
-    if ((fifo_size = fcntl(fd, F_GETPIPE_SZ, 0)) == -1) {
+    if ((fifo_size = fcntl(fd, FD_SETSIZE, 0)) == -1) {
         snprintf(str, DLT_DAEMON_TEXTBUFSIZE, "get FIFO size error: %s\n", strerror(errno));
         dlt_log(LOG_ERR, str);
     }
diff --git a/src/lib/dlt_user.c b/src/lib/dlt_user.c
index cb7f570..720558c 100644
--- a/src/lib/dlt_user.c
+++ b/src/lib/dlt_user.c
@@ -27,6 +27,9 @@
  * \file dlt_user.c
  */
 
+#define PTHREAD_MUTEX_ROBUST  PTHREAD_MUTEX_DEFAULT 
+#define pthread_mutex_consistent pthread_mutex_lock
+
 #include <stdlib.h> /* for getenv(), free(), atexit() */
 #include <string.h> /* for strcmp(), strncmp(), strlen(), memset(), memcpy() */
 #include <signal.h> /* for signal(), SIGPIPE, SIG_IGN */
@@ -413,10 +416,10 @@ DltReturnValue dlt_init(void)
     }
 
     /* make mutex robust to prevent from deadlock when the segmented thread was cancelled, but held the mutex */
-    if (pthread_mutexattr_setrobust(&attr, PTHREAD_MUTEX_ROBUST) != 0) {
-        dlt_user_initialised = false;
-        return DLT_RETURN_ERROR;
-    }
+    //if (pthread_mutexattr_setrobust(&attr, PTHREAD_MUTEX_ROBUST) != 0) {
+    //    dlt_user_initialised = false;
+    //    return DLT_RETURN_ERROR;
+    //}
 
     pthread_mutex_init(&mq_mutex, &attr);
     pthread_mutexattr_destroy(&attr);

**Comment by votrungchi** [comment]
@chenlimingcn You can edit your comment above to put the diff in the diff code block. 
Example: 
<pre>
```diff
- removed
+ added
```
</pre>

**Comment by ssugiura** [comment]
@chenlimingcn Thanks, great to know that you could compile successfully for Windows too. As suggested from @votrungchi , please modify your comment to show the code diff properly. It's even worth to write it down somewhere in the doc.

**Comment by chenlimingcn** [comment]
@ssugiura have done as above

**Comment by manojc99** [comment]
Awesome, Will try this and update

**Comment by koreyoshi-wang** [comment]
I have compiled it with cygwin before and it works, if you need I can share the compiled file to you.

**Comment by ssugiura** [comment]
@koreyoshi-wang , Have you done with cygwin without the change proposed by @chenlimingcn , or any other steps required?
Now I know that there are couple of use cases where users want to use DLT on Windows, so I would like to collect all of suggestions and take them into account.

**Comment by ssugiura** [comment]
@koreyoshi-wang This is a friendly reminder. Please let us know how you compiled for Windows.

**Comment by manojc99** [comment]
Hi @chenlimingcn , I got the below error while doing make install on cygwin. 

$ make install
[  1%] Linking C shared library cygdlt-2.dll
/usr/lib/gcc/x86_64-pc-cygwin/7.4.0/../../../../x86_64-pc-cygwin/bin/ld: cannot find -lsocket
collect2: error: ld returned 1 exit status
make[2]: *** [src/lib/CMakeFiles/dlt.dir/build.make:174: src/lib/cygdlt-2.dll] Error 1
make[1]: *** [CMakeFiles/Makefile2:160: src/lib/CMakeFiles/dlt.dir/all] Error 2
make: *** [Makefile:130: all] Error 2

**Comment by koreyoshi-wang** [comment]
> @koreyoshi-wang This is a friendly reminder. Please let us know how you compiled for Windows.

Hi ssugiura, I have replaced a dependence (pthreads
[pthreads-w32-2-9-1-release.zip](https://github.com/GENIVI/dlt-daemon/files/3879027/pthreads-w32-2-9-1-release.zip)
 for cygwin, and no  change proposed by @chenlimingcn.

**Comment by manojc99** [comment]
@koreyoshi-wang can you please share the precompiled binaries

**Comment by koreyoshi-wang** [comment]
> @koreyoshi-wang can you please share the precompiled binaries

[dlt.zip](https://github.com/GENIVI/dlt-daemon/files/3886176/dlt.zip)
It's compiled with 64-bit system, and tested on windows 10.
Filter didn't work, and I have tried with Linux it didn't work too, so I don't know why.

**Comment by senthil4321** [comment]
@koreyoshi-wang Thank you for sharing the binary. Could you able to fix filter issue? Can you please share build instructions for windows?

**Comment by ssugiura** [comment]
@koreyoshi-wang @senthil4321 Any update on this?

**Comment by jayy89** [comment]
@koreyoshi-wang Could you please let me know a way to replace pthreads dependency for cygwin?

**Comment by koreyoshi-wang** [comment]
> @koreyoshi-wang @senthil4321 Any update on this?

No, I just compiled once, and filter didn't work, but I tested Linux Version the same problem, I don not know the recent status.

**Comment by koreyoshi-wang** [comment]
> @koreyoshi-wang Could you please let me know a way to replace pthreads dependency for cygwin?

Sorry, I have uninstalled my cygwin, you just need to search pthreads in your cygwin library and replace ti.

**Comment by senthil4321** [comment]
I managed to make the filter work. Some points to note. 1.  Filter should be in xml format.(Used dlt viewer gui application to create filter) 2. Both Appication ID and Context ID needs to be enabled in the filter.

**Comment by cogle** [comment]
@manojc99 
To avoid the socket error comment out all instances of:

```
    #set(RT_LIBRARY "")
    #set(SOCKET_LIBRARY socket)
```
and change to(Note these can be copied directly from the if statement above.)
```
    set(RT_LIBRARY rt)
    set(SOCKET_LIBRARY "")
```

The cygwin environment does not have libsocket, but should have `librt`

**Comment by iipeace** [comment]
may I get libdlt.dll that exported all symbols?

**Comment by cogle** [comment]
If you are asking me; I no longer have access to the build.

**Comment by Nikhil-V-maker** [comment]
> > @koreyoshi-wang can you please share the precompiled binaries
> 
> [dlt.zip](https://github.com/GENIVI/dlt-daemon/files/3886176/dlt.zip) It's compiled with 64-bit system, and tested on windows 10. Filter didn't work, and I have tried with Linux it didn't work too, so I don't know why.

Hi @koreyoshi-wang could you share info about how you compiled it. I want to compile the latest version i.e. 2.18.8.

**Comment by koreyoshi-wang** [comment]
Found this repository which has pre-compile file https://github.com/mbehr1/adlt

**Comment by minminlittleshrimp** [comment]
Hello @michael-methner 
I find this topic DLT on Window very interesting.
Should we have an official doc for this topic?
If it is OK, I will collect and summarize all users' idea or methods and prepare for a PR.
Regards

**Comment by thanhthieuBKDN** [comment]
I'm building a C++ app on Window by Visual Studio. I also finish installing DLT daemon by Cygwin on window.
How can we send a log message from C++ app to DLT daemon by using DLT library? Is it possible on Window?

**Comment by minminlittleshrimp** [comment]
Hello @thanhthieuBKDN 
Not sure about dlt on Window since I have not tried yet.
For dlt usage you can refer to this doc: https://github.com/COVESA/dlt-daemon/blob/master/doc/dlt_for_developers.md
Regards

**Comment by humbicho** [comment]
@koreyoshi-wang hi, is there any way to set the value of the port in dlt-receive.exe?
or is there any other way to do it?
